### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by andiwand**
- out source xml translation logic
- out source common functions
- adding abstractions for path, stream and storage
- major clean up and restructure

see https://github.com/andiwand/OpenDocument.core/issues/25
depends on #33

### CODE REVIEW

File: src/TranslationHelper.cpp [hunk 1]
```diff
@@ -1,42 +1,90 @@
...
 
-    TranslationHelperImpl() {
-        translator = OpenDocumentTranslator::create();
-        translatorMs = MicrosoftTranslator::create();
+    TranslationContext context;
+
+    OpenDocumentTranslator translatorOd;
+    MicrosoftTranslator translatorMs;
+
+    Impl() {
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: constructor could be omitted

File: src/TranslationHelper.cpp [hunk 2]
```diff
@@ -1,42 +1,90 @@
...
-        if (!file.isOpen() || !file.isDecrypted()) {
+        if (fileOd.isOpen()) {
+            return translateOd(out, config);
+        }
+        return translateMs(out, config);
+    }
+
+    bool translateOd(const std::string &out, const TranslationConfig &config) {
+        if (!fileOd.isOpen() || !fileOd.isDecrypted()) {
             return translateMs(out, config);
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: wrong

File: src/TranslationHelper.cpp [hunk 3]
```diff
@@ -1,42 +1,90 @@
...
     bool translate(const std::string &out, const TranslationConfig &config) {
-        if (!file.isOpen() || !file.isDecrypted()) {
+        if (fileOd.isOpen()) {
+            return translateOd(out, config);
+        }
+        return translateMs(out, config);
+    }
+
+    bool translateOd(const std::string &out, const TranslationConfig &config) {
+        if (!fileOd.isOpen() || !fileOd.isDecrypted()) {
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: remove is open check

File: src/TranslationHelper.cpp [hunk 4]
```diff
@@ -1,42 +1,90 @@
...
             return translateMs(out, config);
         }
 
-        TranslationContext context = {};
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: keep context reset

File: src/TranslationHelper.cpp [hunk 5]
```diff
@@ -47,49 +95,93 @@ class TranslationHelperImpl final {
             return false;
         }
 
-        TranslationContext context = {};
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: keep context reset

File: src/TranslationHelper.cpp [hunk 6]
```diff
@@ -47,49 +95,93 @@ class TranslationHelperImpl final {
...
         context.config = &config;
         context.msFile = &fileMs;
         context.meta = &fileMs.getMeta();
+        context.currentTextTranslationIndex = 0;
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: remove

File: src/TranslationHelper.cpp [hunk 7]
```diff
@@ -47,49 +95,93 @@ class TranslationHelperImpl final {
...
+            default:
+                return false;
+        }
+    }
+
+    bool backTranslate(const std::string &in, const std::string &out) {
+        if (fileOd.isOpen()) {
+            return backTranslateOd(in, out);
+        }
+        return backTranslateMs(in, out);
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: check open

File: src/TranslationHelper.cpp [hunk 8]
```diff
@@ -1,42 +1,90 @@
...
+        }
+        return fileMs.getMeta();
     }
 
     bool translate(const std::string &out, const TranslationConfig &config) {
-        if (!file.isOpen() || !file.isDecrypted()) {
+        if (fileOd.isOpen()) {
+            return translateOd(out, config);
+        }
+        return translateMs(out, config);
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: check open

File: src/TranslationHelper.cpp [hunk 9]
```diff
@@ -47,49 +95,93 @@ class TranslationHelperImpl final {
...
+
+    bool backTranslate(const std::string &in, const std::string &out) {
+        if (fileOd.isOpen()) {
+            return backTranslateOd(in, out);
+        }
+        return backTranslateMs(in, out);
+    }
+
+    bool backTranslateOd(const std::string &in, const std::string &out) {
+        if (!fileOd.isOpen() || !context.config->editable) {
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: remove isOpen

File: src/TranslationHelper.cpp [hunk 10]
```diff
@@ -47,49 +95,93 @@ class TranslationHelperImpl final {
...
+            case FileType::OPENDOCUMENT_SPREADSHEET:
+            case FileType::OPENDOCUMENT_GRAPHICS:
+                return translatorOd.backTranslate(fileOd, in, out, context);
+            default:
+                return false;
+        }
+    }
+
+    bool backTranslateMs(const std::string &in, const std::string &out) {
+        if (!fileMs.isOpen() || !context.config->editable) {
```
Comment (andiwand) [inline src/TranslationHelper.cpp]: remove isOpen

File: src/TranslationHelper.cpp (hunk 5 above)
Comment (andiwand) [inline src/TranslationHelper.cpp]: remove isOpen

File: src/io/FileUtil.h [hunk 11]
```diff
@@ -0,0 +1,14 @@
+#ifndef ODR_FILEUTIL_H
+#define ODR_FILEUTIL_H
+
+#include <string>
+
+namespace odr {
+
+namespace FileUtil {
+std::string read(const std::string &);
```
Comment (andiwand) [inline src/io/FileUtil.h]: extern

**Comment by andiwand** [comment]
TODO fix ods editing
TODO split pr in refactor / editing / xlsx

File: src/ms/MicrosoftTranslator.cpp [hunk 12]
```diff
@@ -69,6 +69,56 @@ class MicrosoftTranslator::Impl {
 
         contentTranslator.translate(*body, context);
     }
+
+    bool backTranslate(MicrosoftOpenXmlFile &in, const std::string &inHtml, const std::string &out, TranslationContext &context) const {
+        // TODO code duplication
+        // TODO exit on encrypted files
```
Comment (TomTasche) [inline src/ms/MicrosoftTranslator.cpp]: also doesn't work in java at the moment

Comment (andiwand) [reply src/ms/MicrosoftTranslator.cpp]: what does not work?

Comment (TomTasche) [reply src/ms/MicrosoftTranslator.cpp]: saving an encrypted document after edit

Comment (andiwand) [reply src/ms/MicrosoftTranslator.cpp]: that's on my list. i would like to have html / css templates as resources.

**Pull request merged**